### PR TITLE
[CALCITE-6396] Add ADD_MONTHS function (enabled in Spark library)

### PR DESCRIPTION
**Pull request by caicancai**
https://issues.apache.org/jira/browse/CALCITE-6396

### CODE REVIEW

File: site/_docs/reference.md [hunk 1]
```diff
@@ -2674,6 +2674,7 @@ In the following:
 | p | expr :: type                                   | Casts *expr* to *type*
 | m | expr1 <=> expr2                                | Whether two values are equal, treating null values as the same, and it's similar to `IS NOT DISTINCT FROM`
 | * | ACOSH(numeric)                                 | Returns the inverse hyperbolic cosine of *numeric*
+| s | ADD_MONTHS(start_date, num_months)             | Returns the date that is *num_months* after *start_date*.
```
Comment (mihaibudiu) [inline site/_docs/reference.md:2677]: does this operate on dates or strings?
the code assumes strings

Comment (caicancai) [reply site/_docs/reference.md:2677]: strings

Comment (caicancai) [reply site/_docs/reference.md]: @mihaibudiu I don't know why cli reports an error

Comment (mihaibudiu) [reply site/_docs/reference.md]: what cli?

Comment (caicancai) [reply site/_docs/reference.md]: 2024-05-07 08-30-46屏幕截图

Comment (mihaibudiu) [reply site/_docs/reference.md]: that is not a CLI, that is the CI

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java [hunk 2]
```diff
@@ -270,6 +270,13 @@ private static SqlCall transformConvert(SqlValidator validator, SqlCall call) {
               .andThen(SqlTypeTransforms.TO_NULLABLE_ALL),
           OperandTypes.SAME_SAME);
 
+  /** The "NVL(value, value)" function. */
+  @LibraryOperator(libraries = {SPARK})
+  public static final SqlBasicFunction ADD_MONTHS =
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:275]: According to the Databricks documentation ADD_MONTHS takes an expression that evaluates to a date, and not a string: https://docs.databricks.com/en/sql/language-manual/functions/add_months.html

The result is also supposed to be a DATE.

I have already asked this question in a prior review. Reviews take a lot of energy and time, so please listen to the comments if you expect people to review your PRs.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 3]
```diff
@@ -10418,6 +10418,23 @@ void checkNvl(SqlOperatorFixture f0, FunctionAlias functionAlias) {
...
 
+  @Test void testAddMonthsFunc() {
+    final SqlOperatorFixture f0 = fixture();
+    f0.setFor(SqlLibraryOperators.ADD_MONTHS);
+    f0.checkFails("^add_months('2016-08-31', 1)^",
+        "No match found for function signature "
+            + "ADD_MONTHS\\(<CHARACTER>, <NUMERIC>\\)", false);
+
+    final SqlOperatorFixture f = f0.withLibrary(SqlLibrary.SPARK);
+    f.checkScalar("ADD_MONTHS('2016-08-31', 1)", "2016-09-30",
```
Comment (mihaibudiu) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:10429]: you should test with invalid dates too, and with dates prior to the Gregorian calendar introduction.
You should also test with negative values for the number of months.

**Review by mihaibudiu — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java (hunk 2 above)
Comment (caicancai) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:275]: Thanks for reminding me, I will improve this PR

Comment (caicancai) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:275]: Thank you

Comment (caicancai) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:275]: I encountered a weird bug that seems to take some time

**Pull request closed without merging**